### PR TITLE
raftstore/store: take snapshots in apply worker

### PR DESCRIPTION
**Pull request by overvenus**
## What have you changed? (mandatory)

This PR change snapshot procedure, it now takes snapshot of raft engine and kv engine in the apply worker instead of the region worker. Before taking snapshot, we sync two engines first.

This is splited from #4372 .  

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

Unit tests and integration tests.

## Does this PR affect documentation (docs) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

#4372

### CODE REVIEW

File: src/raftstore/store/fsm/apply.rs [hunk 1]
```diff
@@ -2514,6 +2526,20 @@ impl ApplyFsm {
...
+    fn handle_snapshot(&mut self, apply_ctx: &mut ApplyContext, snap_task: GenSnapTask) {
+        if self.delegate.pending_remove || self.delegate.stopped {
+            return;
+        }
+
+        apply_ctx.prepare_for(&self.delegate);
+        // persistent all pending changes
+        apply_ctx.commit_opt(&mut self.delegate, true);
+        if let Err(e) = snap_task.generate_snapshot(&apply_ctx.engines) {
+            error!("schedule snapshot failed"; "error" => ?e);
```
Comment (hicqu) [inline src/raftstore/store/fsm/apply.rs]: Can `region_id` and `peer_id` be contained?

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -1137,14 +1164,20 @@ impl PeerStorage {
     }
 
     /// Update the memory state after ready changes are flushed to disk successfully.
-    pub fn post_ready(&mut self, ctx: InvokeContext) -> Option<ApplySnapResult> {
+    pub fn post_ready(
+        &mut self,
+        ctx: InvokeContext,
+    ) -> (Option<ApplySnapResult>, Option<GenSnapTask>) {
+        // The snapshot gen task need to be sent the apply system.
+        let gen_snap = self.snap_gen_task.borrow_mut().take();
```
Comment (hicqu) [inline src/raftstore/store/peer_storage.rs]: What if the region hibernate? Followers can request snapshots even if there are no proposals for the region.

Comment (overvenus) [reply src/raftstore/store/peer_storage.rs]: I think it will be invoked as long as follower requests snapshot via raft message.

**Comment by hicqu** [comment]
LGTM.

File: src/raftstore/store/fsm/apply.rs [hunk 3]
```diff
@@ -2514,6 +2562,27 @@ impl ApplyFsm {
...
     }
 
+    fn handle_snapshot(&mut self, apply_ctx: &mut ApplyContext, snap_task: GenSnapTask) {
+        if self.delegate.pending_remove || self.delegate.stopped {
+            return;
+        }
+
+        apply_ctx.prepare_for(&self.delegate);
+        // persistent all pending changes
+        apply_ctx.commit_opt(&mut self.delegate, true);
```
Comment (BusyJay) [inline src/raftstore/store/fsm/apply.rs]: Why not just flush?

**Pull request merged**